foyer/xml_writer.py

```diff
@@ -108,6 +110,37 @@ def _write_atoms(self, root, atoms, forcefield, unique):
         nb_force.set('sigma', str(round(atom.atom_type.sigma/10, 4)))
         nb_force.set('epsilon', str(round(atom.atom_type.epsilon * 4.184, 6)))
 
+    _update_defs(atomtypes, nonbonded, forcefield)
+
+def _update_defs(atomtypes, nonbonded, forcefield):
+    def_list = [i.get('def') for i in atomtypes.iterchildren()]
+    name_list = [i.get('name') for i in atomtypes.iterchildren()]
+    smarts_list = list()
+    smarts_parser = forcefield.parser
+    for smarts_string, name in zip(def_list, name_list):
+        smarts_graph = SMARTSGraph(smarts_string, parser=smarts_parser,
+                                   name=name)
+        for atom_expr in nx.get_node_attributes(smarts_graph, name='atom').values():
+            labels = atom_expr.find_data('has_label')
+            for label in labels:
+                atom_type = label.children[0][1:]
+                smarts_list.append(atom_type)
+    smarts_list = list(set(smarts_list))
+    extra_types = [i for i in smarts_list if i not in name_list]
+
+    for extra in extra_types:
```

Comment:
So in this nested-for-loop, it looks like you're looking at all the `extra_types` (superfluous atomtypes that were mentioned in smarts strings) and finding the smarts definitions that contain these extra atomtypes, then formatting the smarts definition to exclude the extra atomtype. If you found an extra atomtype in a smarts definition, then you iterate through ALL your xml-atomtype-elements to replace/modify the smarts definition.

I think one point of concern is that if you find ONE smarts definition that contains the extra type, you then modify ANY smarts definitions that contain the extra type with this new-definition, not just modifying that particular smarts definition. I think this will only work if all smarts definitions containing "extra" end up using the exact same smarts definition. 

Could you do something like:
```suggestion
    for extra in extra_types:
        for i, definition in enumerate(def_list):
            if extra in definition:
                extra_edit = '%' + extra
                extra_index = definition.find(extra_edit)
                if definition[extra_index-1] == ';':
                    new_def = definition.replace(extra_edit + ',' , '')
                else:
                    new_def = definition.replace(',' + extra_edit, '')
                atomtypes[i].set('def', new_def)
                

```
This way, you only set one smarts definition as `new_def` and not any smarts definition with `extra` inside the definition

Reply:
I like this method better.